File: src/ingestion/__fixtures__/python/services.py
from typing import Optional, List


class UserRepository:
    def find_by_id(self, user_id: int) -> Optional["User"]:
        return db.query(User).filter(User.id == user_id).first()

    def find_by_email(self, email: str) -> Optional["User"]:
        return db.query(User).filter(User.email == email).first()

    def save(self, user: "User") -> "User":
        db.add(user)
        db.commit()
        return user


class EmailService:
    def send_welcome(self, email: str, name: str) -> bool:
        template = self.render_template("welcome", name=name)
        return self.send(email, "Welcome!", template)

    def send_reset_password(self, email: str, token: str) -> bool:
        template = self.render_template("reset", token=token)
        return self.send(email, "Reset your password", template)

    def send(self, to: str, subject: str, body: str) -> bool:
        pass

    def render_template(self, name: str, **kwargs) -> str:
        pass


class PaymentGateway:
    def charge(self, amount: float, currency: str, token: str) -> "PaymentResult":
        pass

    def refund(self, transaction_id: str) -> "RefundResult":
        pass

    def get_balance(self) -> float:
        pass

File: src/ingestion/__fixtures__/rust/services.rs
use std::collections::HashMap;

pub struct AuthService {
    secret_key: String,
}

impl AuthService {
    pub fn verify(&self, token: &str) -> Result<Claims, AuthError> {
        jwt::decode(token, &self.secret_key)
    }

    pub fn create_token(&self, user_id: u64) -> String {
        jwt::encode(user_id, &self.secret_key)
    }

    pub fn revoke_token(&self, token: &str) -> Result<(), AuthError> {
        self.blacklist.insert(token.to_string());
        Ok(())
    }
}

pub struct DatabasePool {
    pool: Vec<Connection>,
}

impl DatabasePool {
    pub fn query(&self, sql: &str, params: &[&str]) -> Result<Vec<Row>, DbError> {
        let conn = self.pool.first().unwrap();
        conn.execute(sql, params)
    }

    pub fn execute(&self, sql: &str, params: &[&str]) -> Result<u64, DbError> {
        let conn = self.pool.first().unwrap();
        conn.execute_mut(sql, params)
    }
}

pub struct MailService {
    smtp_host: String,
}

impl MailService {
    pub fn send_welcome(&self, email: &str) -> Result<(), MailError> {
        let body = format!("Welcome to our platform!");
        self.send(email, "Welcome!", &body)
    }

    pub fn send_reset(&self, email: &str, token: &str) -> Result<(), MailError> {
        let body = format!("Reset link: /reset?token={}", token);
        self.send(email, "Password Reset", &body)
    }

    fn send(&self, to: &str, subject: &str, body: &str) -> Result<(), MailError> {
        Ok(())
    }
}
